File: solutions/cpp/Task_assignment.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <iostream>
#include <limits>
#include <random>
#include <utility>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::numeric_limits;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

// @include
struct PairedTasks {
  int task_1, task_2;
};

vector<PairedTasks> OptimumTaskAssignment(vector<int> task_durations) {
  sort(task_durations.begin(), task_durations.end());
  vector<PairedTasks> optimum_assignments;
  for (int i = 0, j = task_durations.size() - 1; i < j; ++i, --j) {
    optimum_assignments.emplace_back(
        PairedTasks{task_durations[i], task_durations[j]});
  }
  return optimum_assignments;
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = atoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 10000);
    n = dis(gen);
  }
  vector<int> A;
  for (size_t i = 0; i < n; ++i) {
    uniform_int_distribution<int> dis(0, 999);
    A.emplace_back(dis(gen));
  }
  vector<PairedTasks> P(OptimumTaskAssignment(A));
  int max = numeric_limits<int>::min();
  for (size_t i = 0; i < P.size(); ++i) {
    if (P[i].task_1 + P[i].task_2 > max) {
      max = P[i].task_1 + P[i].task_2;
    }
  }
  cout << max << endl;
  return 0;
}

File: static/cpp/java/src/main/java/com/epi/TaskAssignment.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class TaskAssignment {
  // @include
  private static class PairedTasks {
    public Integer task1;
    public Integer task2;

    public PairedTasks(Integer task1, Integer task2) {
      this.task1 = task1;
      this.task2 = task2;
    }
  }

  public static List<PairedTasks> optimumTaskAssignment(
      List<Integer> taskDurations) {
    Collections.sort(taskDurations);
    List<PairedTasks> optimumAssignments = new ArrayList<>();
    for (int i = 0, j = taskDurations.size() - 1; i < j; ++i, --j) {
      optimumAssignments.add(
          new PairedTasks(taskDurations.get(i), taskDurations.get(j)));
    }
    return optimumAssignments;
  }
  // @exclude

  public static void main(String[] args) {
    Random gen = new Random();
    int n;
    if (args.length == 1) {
      n = Integer.parseInt(args[0]);
    } else {
      n = gen.nextInt(10000) + 1;
    }
    List<Integer> A = new ArrayList<>(n);
    for (int i = 0; i < n; ++i) {
      A.add(gen.nextInt(999));
    }
    List<PairedTasks> P = optimumTaskAssignment(A);
    int max = Integer.MIN_VALUE;
    for (PairedTasks aP : P) {
      if (aP.task1 + aP.task2 > max) {
        max = aP.task1 + aP.task2;
      }
    }
    System.out.println(max);
  }
}
